operator/v1/0000_70_cluster-network-operator_01.crd.yaml

```diff
@@ -290,9 +290,23 @@ spec:
                         type: integer
                     type: object
                   ovnKubernetesConfig:
-                    description: oVNKubernetesConfig configures the ovn-kubernetes
-                      plugin. This is currently not implemented.
+                    description: ovnKubernetesConfig configures the ovn-kubernetes
+                      plugin.
                     properties:
+                      gatewayConfig:
+                        description: gatewayConfig holds the configuration for node
+                          gateway options. Supported from OCP 4.10.
+                        properties:
+                          routingViaHost:
+                            default: false
+                            description: RoutingViaHost allows pod egress traffic
+                              to exit via the ovn-k8s-mp0 management port into the
+                              host before sending it out. If this is not set, traffic
+                              will always egress directly from OVN to outside without
+                              touching the host stack. Setting this to true means
+                              hardware offload will not be supported. Default is false.
```

Comment:
What do we do when we upgrade existing clusters?  Do we need to handle that case specifically?

Reply:
So upgrades are a bit different, currently with this feature we'll only be supporting 4.9 -> 4.10 cases. Since we kind of stopped supporting lgw around 4.8, in 4.8/4.9 lgw is currently unsupported - same for the upgrades.

But with this bool, the aim is to allow users to be able to switch between both the modes at any given time - i.e let's say you are on 4.9 & shared gateway mode. If you want to upgrade to 4.10 and choose lgw mode, you'd have to first do the upgrade to 4.10 shared and then edit this bool in the crd and CNO will roll out ovnkube pods in local gateway. Vice versa if you are coming from 4.9 local gateway and you want to move to 4.10 shared you'd have to first go to 4.10 local and switch off this bool (which it is by default, so by default we are always shared). So essentially users should be able to go between the two modes at any given point since the --to-upgrade doesn't really provide a way to specific like get me directly to x gateway mode.

Reply:
Another point to note is, AFAIK, the api is the first thing to upgrade and cno is the third component, so scenario:

We are moving from 4.9 local to 4.10 local:

- If the API upgrades first, the CRD changes for the `gatewayconfig` should be available in the client-go for the user to do an `oc edit` (I think). So API is already on 4.10.
- At this point CNO is still on 4.9, and CNO is the last thing to upgrade in the sense it rolls out all multus/ovn-k pods and then it itself undergoes upgrade, so it rolls out all the ovnkube daemonsets from 4.9 to 4.10 and since the 4.9 CNO used the hidden config-map to set the gateway mode, it would roll out the new ones respecting that config-map.mode value.
- Once all ovn-k pods are on 4.10, CNO also upgrades, the new CNO now recognizes the new gatewayconfig values if they are set.

I need to test this process when this bool is merged and basically make CNO control the mode migration process during upgrades.

Reply:
I guess my question is whether we should say something like "the default is false unless upgrading from an earlier version"?

Reply:
ah :D you make a good point :) . yes we definitely need to respect the mode that existed in the previous version of the cluster. Let me update that.